index.d.ts

```diff
@@ -0,0 +1,226 @@
+    }
+    export class MutableFile extends File {
+        storage: Storage;
+        static packAttributes(attributes: Object): Buffer;
+        constructor(opts: FileOpts, storage: Storage);
+        mkdir(opts: mkdirOpts | string, cb?: (error: err, file: Nullable<MutableFile>) => void): void;
+        upload(opts: uploadOpts | string, source?: BufferString, cb?: uploadCb): Writable;
+        uploadAttribute(type: 0 | 1, data: Buffer, callback?: (error: err, file?: this) => void): void;
+        // Not sure about type of file in delete's cb
+        delete(permanent?: boolean, cb?: (error: err, file?: File) => void): this;
+        moveTo(target: File | string, cb?: (error: err, file?: File) => void): this;
+        setAttributes(attributes: Object, cb?: noop): this;
+        rename(filename: string, cb?: noop): this;
+        setLabel(label: labelType, cb?: noop): this;
+        setFavorite(isFavorite?: boolean, cb?: noop): this;
+        shareFolder(options: linkOpts, cb?: noop): this;
+        // this method has a option paramteter but never uses it, maybe update this in the file or change it here 
+        unshareFolder(cb?: noop): this;
+        importFile(sharedFile: string | File, cb?: (error: err, file?: this) => void): void | this;
+        on(event: 'move', listener: (oldDir: File) => void): this;
+        on(event: 'update', listener: (file: MutableFile) => void): this;
+        on(event: 'delete', listener: (file: Readonly<File>) => void): this;
+        once(event: 'move', listener: (oldDir: File) => void): this;
+        once(event: 'update', listener: (file: MutableFile) => void): this;
+        once(event: 'delete', listener: (file: Readonly<File>) => void): this;
+    }
+    export class AES {
+        key: Buffer;
+        constructor(key: Buffer);
+        encryptCBC(buffer: Buffer): Buffer;
+        decryptCBC(buffer: Buffer): Buffer;
+        stringhash(buffer: Buffer): Buffer;
+        encryptECB(buffer: Buffer): Buffer;
+        decryptECB(buffer: Buffer): Buffer;
+    }
+    // Interfaces & Types
+    type labelType = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | '' | 'red' | 'orange' | 'yellow' | 'green' | 'blue' | 'purple' | 'grey';
+    type StorageStatus = 'ready' | 'connecting' | 'closed';
+    type uploadCb = (error: err, file: MutableFile) => void;
+    type errorCb = (error: err) => void;
+    type BufferString = Buffer | string;
+    type Nullable<T> = T | null;
+    type err = Nullable<Error>;
+    type noop = () => void;
+    type Fetch = any; // Change this if you can get the type of fetch
+    interface StorageOpts {
+        email: string;
+        password: string;
+        autoload?: boolean;
+        autologin?: boolean;
+        keepalive?: boolean;
+    }
+    interface StorageJSON {
+        key: string;
+        sid: string;
+        name: string;
+        user: any; // Not sure what this is
+        options: StorageOpts;
+    }
+    interface APIOpts {
```

Comment:
I don't know but in the API class's constructor only `gateway` & `fetch` options were considered.

Reply:
btw node-fetch does not support `credentials` in request option ([Ref](https://github.com/node-fetch/node-fetch/tree/2.x#class-request)) which is being passed in by `defaultFetch` method in [API class](https://github.com/qgustavor/mega/blob/replacing-request/lib/api.mjs#L56). I thought I should tell you this

Reply:
> I don't know but in the API class's constructor only `gateway` & `fetch` options were considered so I only added those two to the interface.

That's true, for some reason I forgot to add `this.userAgent = opt.userAgent || ...`. I will fix that.

> btw node-fetch does not support `credentials` in request option ([Ref](https://github.com/node-fetch/node-fetch/tree/2.x#class-request)) which is being passed in by `defaultFetch` method in [API class](https://github.com/qgustavor/mega/blob/replacing-request/lib/api.mjs#L56). I thought I should tell you this

That's fine, the credentials option is not meant for node-fetch but for browser: without it the library don't works in the browser because fetch throws as user-agent is not an allowed header.

Reply:
About the credentials option: I tested again the library and in Chrome setting it stops the headers issue but in Firefox this issue continues. What to do? Make a Deno build?

Reply:
I will remove the `same-origin` option as it don't solves the issue and is the default value since 2018. If needed I will make a Deno build, but I'll look for workarounds first. The good thing is that when Node implements fetch then the Deno build will probably work in Node (except by the PBKDF2 part, but that's easy to fix).

Reply:
I made so if `userAgent` is `null` it would not set it. I need to add this to the documentation: "If you're targeting browsers (mostly Firefox) then set `userAgent` to `null`.)